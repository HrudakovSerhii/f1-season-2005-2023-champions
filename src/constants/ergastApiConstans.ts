export const ROOT_ERGAST_API_PATH = 'https://ergast.com/api/f1';

export const ALL_SEASON_WINNERS_PATH = '/driverStandings/1.json';

export const SEASON_WINNERS_PATH = '/{season}/results/1.json';