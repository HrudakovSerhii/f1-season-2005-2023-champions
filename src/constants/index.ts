export const isDevelopment = process.env.NODE_ENV !== 'production'

export const isTest = process.env.NODE_ENV === 'test'

export const SEASONS_RANGE: [number, number] = [2005, 2023]

export const MONTH_NAMES = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
]

// App routing
export const HOME = '/'
export const SEASON = '/season/:seasonId'
export const SEASONS = '/seasons'

export const ABOUT = '/about'
export const CONTACT = '/contact'

export const CONTACT_EMAIL = 'hrudakovserhii@gmail.com'
