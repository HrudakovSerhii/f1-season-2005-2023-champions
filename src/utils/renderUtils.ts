import { StandingItem } from '../types'

export const getSeasonsWinnersGroupedData = (
    data: StandingItem[],
    groupSize: number = 5
) =>
    data?.reduce((groupedItems, standingItem, index) => {
        const groupIndex = Math.floor(index / groupSize)

        if (groupedItems[groupIndex]) {
            groupedItems[groupIndex] = [
                ...groupedItems[groupIndex],
                data[index],
            ]
        } else {
            groupedItems.push([data[index]])
        }

        return groupedItems
    }, [] as StandingItem[][])
