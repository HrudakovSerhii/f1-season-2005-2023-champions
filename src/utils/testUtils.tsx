import React from 'react'
import { Router } from 'react-router'

import { render, RenderOptions } from '@testing-library/react'

/**
 * Utility function to wrap component in React Router
 * @param ui - component which consist react router elements in render tree
 * @param history - memory History
 * @param options - @testing-library/react render options
 */
export const renderWithRouter = (
    ui: React.ReactElement,
    history: any,
    options?: Omit<RenderOptions, 'queries'>
) => {
    const Wrapper: React.FC<{ children: React.ReactElement }> = ({
        children,
    }) => (
        <Router location={history.location} navigator={history}>
            {children}
        </Router>
    )

    return render(ui, { wrapper: Wrapper, ...options })
}
