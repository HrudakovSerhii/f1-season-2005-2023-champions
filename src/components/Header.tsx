import React from 'react'
import { Link } from 'react-router-dom'

import { ABOUT, CONTACT, HOME, SEASONS, SEASONS_RANGE } from '../constants'

const Header = () => {
    return (
        <header className="app-header">
            <nav className="border-b-2 border-gray-300">
                <div className="flex flex-wrap items-center justify-between p-4">
                    <a
                        href="https://www.formula1.com/"
                        className="flex items-center"
                    >
                        <img
                            src="/assets/f1-icon.png"
                            className="h-12 mr-1"
                            alt="F1 Logo"
                        />
                        <div className="flex flex-col">
                            <span className="self-center text-2xl font-semibold whitespace-nowrap">
                                Champions
                            </span>
                            <span className="text-xs">
                                Season {SEASONS_RANGE[0]}-{SEASONS_RANGE[1]}
                            </span>
                        </div>
                    </a>
                    <div className="w-auto">
                        <ul className="font-medium text-lg flex flex-row space-x-8">
                            <li>
                                <Link
                                    to={HOME}
                                    className="hover:text-blue-700"
                                    aria-current="page"
                                    title="Go to Home page"
                                >
                                    Home
                                </Link>
                            </li>
                            <li>
                                <Link
                                    to={SEASONS}
                                    className="hover:text-blue-700"
                                    title="Go to Seasons page to view seasons full information"
                                >
                                    Seasons
                                </Link>
                            </li>
                            <li>
                                <Link
                                    to={ABOUT}
                                    className="hover:text-blue-700"
                                    title="Go to About page to see SPA description and targets"
                                >
                                    About
                                </Link>
                            </li>
                            <li>
                                <Link
                                    to={CONTACT}
                                    className="hover:text-blue-700"
                                    title="Go to Contact page to check SPA author information and contacts"
                                >
                                    Contact
                                </Link>
                            </li>
                        </ul>
                    </div>
                </div>
            </nav>
        </header>
    )
}

export default Header
