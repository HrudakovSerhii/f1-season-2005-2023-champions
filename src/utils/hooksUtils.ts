import { StandingItem } from '../types'
import { SEASONS_RANGE } from '../constants'

/**
 * Function filter standingsLists to get winners only for range specified in App requirements.
 * This could be done by using offset param. on StandingItem[] data fetch but wasn't chosen as it will require to fetch
 * current season data first to see current season value.
 * @param data {StandingItem[]} list of season items
 * @param seasonRange {SEASONS_RANGE} tuple of seasons - default to [2005-2023]
 * @return StandingItem[]
 */
export const filterWinnersBySeasonRange = (
    data: StandingItem[],
    seasonRange: typeof SEASONS_RANGE = SEASONS_RANGE
) =>
    data?.filter(
        (standingItem) =>
            Number(standingItem.season) >= seasonRange[0] &&
            Number(standingItem.season) <= seasonRange[1]
    )

/**
 * Function search and return winner data of specific season
 * @param data {StandingItem[]} list of season items
 * @param season {string} id of the season
 * @return StandingItem
 */
export const selectWinnerBySeason = (data: StandingItem[], season: string) =>
    data.find((standingItem) => standingItem.season === season)
