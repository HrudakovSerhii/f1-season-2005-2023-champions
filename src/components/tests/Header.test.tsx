import React from 'react'
import { fireEvent, screen } from '@testing-library/react'

import { createMemoryHistory } from 'history'

import Header from '../Header'

import { renderWithRouter } from '../../utils/testUtils'

import { HOME, SEASONS } from '../../constants'

describe('Header', () => {
    it('clicking Home link navigates to Home page', () => {
        const history = createMemoryHistory()

        renderWithRouter(<Header />, history)

        fireEvent.click(screen.getByText('Home'))

        expect(history.location.pathname).toBe(HOME)
    })

    it('clicking SEASONS link navigates to SEASONS page', async () => {
        const history = createMemoryHistory()

        renderWithRouter(<Header />, history)

        fireEvent.click(screen.getByText('Seasons'))

        expect(history.location.pathname).toBe(SEASONS)
    })

    it('should render correctly', () => {
        const history = createMemoryHistory()

        const { asFragment } = renderWithRouter(<Header />, history)

        expect(asFragment()).toMatchSnapshot()
    })
})
