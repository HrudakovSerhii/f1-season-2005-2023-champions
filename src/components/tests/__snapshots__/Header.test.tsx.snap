// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly 1`] = `
<DocumentFragment>
  <header
    class="app-header"
  >
    <nav
      class="border-b-2 border-gray-300"
    >
      <div
        class="flex flex-wrap items-center justify-between p-4"
      >
        <a
          class="flex items-center"
          href="https://www.formula1.com/"
        >
          <img
            alt="F1 Logo"
            class="h-12 mr-1"
            src="/assets/f1-icon.png"
          />
          <div
            class="flex flex-col"
          >
            <span
              class="self-center text-2xl font-semibold whitespace-nowrap"
            >
              Champions
            </span>
            <span
              class="text-xs"
            >
              Season 2005-2023
            </span>
          </div>
        </a>
        <div
          class="w-auto"
        >
          <ul
            class="font-medium text-lg flex flex-row space-x-8"
          >
            <li>
              <a
                aria-current="page"
                class="hover:text-blue-700"
                href="/"
                title="Go to Home page"
              >
                Home
              </a>
            </li>
            <li>
              <a
                class="hover:text-blue-700"
                href="/seasons"
                title="Go to Seasons page to view seasons full information"
              >
                Seasons
              </a>
            </li>
            <li>
              <a
                class="hover:text-blue-700"
                href="/about"
                title="Go to About page to see SPA description and targets"
              >
                About
              </a>
            </li>
            <li>
              <a
                class="hover:text-blue-700"
                href="/contact"
                title="Go to Contact page to check SPA author information and contacts"
              >
                Contact
              </a>
            </li>
          </ul>
        </div>
      </div>
    </nav>
  </header>
</DocumentFragment>
`;
